scripts/install_python_dev_dependencies.py

```diff
@@ -0,0 +1,129 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Install Python development dependencies."""
+
+from __future__ import annotations
+
+import argparse
+import os
+import subprocess
+import sys
+
+from typing import List, Optional
+
+
+INSTALLATION_TOOL_VERSIONS = {
+    'pip': '22.1.1',
+    'pip-tools': '6.6.2',
+    'setuptools': '58.5.3',
+}
+REQUIREMENTS_DEV_FILE_PATH = 'requirements_dev.in'
+COMPILED_REQUIREMENTS_DEV_FILE_PATH = 'requirements_dev.txt'
+
+
+def assert_in_venv() -> None:
+    """Raise an error if we are not in a virtual environment.
+
+    No error is raised if we are running on GitHub Actions because a
+    virtual environment is unnecessary there.
+    """
+    if 'GITHUB_ACTION' in os.environ:
+        return
+    if sys.prefix == sys.base_prefix:
+        raise AssertionError(
+            'Oppia must be developed within a virtual environment.')
+
+
+def install_installation_tools() -> None:
+    """Install the minimal tooling needed to install dependencies."""
+    for package, version in INSTALLATION_TOOL_VERSIONS.items():
+        # We run pip as a subprocess because importing from the pip
+        # module is not supported:
+        # https://pip.pypa.io/en/stable/user_guide/#using-pip-from-your-program.
+        subprocess.run(
+            [
+                sys.executable, '-m', 'pip', 'install',
+                f'{package}=={version}'
+            ],
+            check=True,
+            encoding='utf-8',
+        )
+
+
+def install_dev_dependencies() -> None:
+    """Install dev dependencies from
+    COMPILED_REQUIREMENTS_DEV_FILE_PATH.
+    """
+    subprocess.run(
+        ['pip-sync', COMPILED_REQUIREMENTS_DEV_FILE_PATH],
+        check=True,
+        encoding='utf-8',
+    )
+
+
+def compile_dev_dependencies() -> bool:
+    """Generate COMPILED_REQUIREMENTS_DEV_FILE_PATH file.
+
+    Returns:
+        bool. Whether the compiled dev requirements file was changed.
+    """
+    with open(
+            COMPILED_REQUIREMENTS_DEV_FILE_PATH, 'r',
+            encoding='utf-8') as f:
+        old_compiled = f.read()
+    subprocess.run(
+        [
+            'pip-compile', REQUIREMENTS_DEV_FILE_PATH, '--output-file',
+            COMPILED_REQUIREMENTS_DEV_FILE_PATH
+        ],
+        check=True,
+        encoding='utf-8',
+    )
+    with open(
+            COMPILED_REQUIREMENTS_DEV_FILE_PATH, 'r',
+            encoding='utf-8') as f:
+        new_compiled = f.read()
+
+    return old_compiled != new_compiled
+
+
+def main(cli_args: Optional[List[str]] = None) -> None:
+    """Install all dev dependencies."""
+    parser = argparse.ArgumentParser(
+        'Install Python development dependencies')
+    parser.add_argument(
+        '--assert_compiled', action='store_true',
+        help='Assert that the dev requirements file is already compiled.')
```

Comment:
We usually define these at the top of the file.

Reply:
done